clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/IngestionBatchProcessor.java

```diff
@@ -81,39 +81,32 @@ KeyLevelLocksManager getLockManager() {
    * When {@link #lockManager} is not null, this function will try to lock all the keys
    * (except Control Messages) passed by the params.
    */
-  public List<ReentrantLock> lockKeys(List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> records) {
+  public Map<ByteArrayKey, ReentrantLock> lockKeys(List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> records) {
     if (lockManager != null) {
-      List<ReentrantLock> locks = new ArrayList<>(records.size());
-
       /**
-       * Collect all the keys and sort them to avoid deadlock.
+       * Need to use a {@link TreeMap} to make sure the locking will be executed in a deterministic order, otherwise
+       * deadlock can happen.
+       * Considering there could be multiple consumers, which are executing this function concurrently, and if they
+       * are trying to lock the same set of keys with different orders, deadlock can happen.
        */
-      List<byte[]> keys = new ArrayList<>(records.size());
-
+      Map<ByteArrayKey, ReentrantLock> keyLockMap =
+          new TreeMap<>((o1, o2) -> ByteUtils.compare(o1.getContent(), o2.getContent()));
       records.forEach(r -> {
         if (!r.getKey().isControlMessage()) {
-          keys.add(r.getKey().getKey());
+          keyLockMap.computeIfAbsent(ByteArrayKey.wrap(r.getKey().getKey()), k -> lockManager.acquireLockByKey(k));
         }
       });
-      Collections.sort(keys, (o1, o2) -> ByteUtils.compare(o1, o2));
-
-      keys.forEach(key -> {
-        ReentrantLock lock = lockManager.acquireLockByKey(ByteArrayKey.wrap(key));
-        locks.add(lock);
-        lock.lock();
-      });
-      return locks;
+      keyLockMap.forEach((k, v) -> v.lock());
+      return keyLockMap;
     }
-    return Collections.emptyList();
+    return Collections.emptyMap();
   }
 
-  public void unlockKeys(List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> records, List<ReentrantLock> locks) {
+  public void unlockKeys(Map<ByteArrayKey, ReentrantLock> keyLockMap) {
     if (lockManager != null) {
-      locks.forEach(lock -> lock.unlock());
-      records.forEach(r -> {
-        if (!r.getKey().isControlMessage()) {
-          lockManager.releaseLock(ByteArrayKey.wrap(r.getKey().getKey()));
-        }
+      keyLockMap.forEach((key, lock) -> {
+        lock.unlock();
+        lockManager.releaseLock(key);
```

Comment:
This is pre-existing code, not something you are changing in this PR, but I wonder... why did we design the `KeyLevelLocksManager` in this way? IIUC, the `lock.unlock()` unlocks (duh) and then the `lockManager.releaseLock(key)` decreases the ref count and potentially returns the lock to the pool... Is there ever a situation where we would wish to do one of these operations but not the other? At first glance, it seems to me that the only correct way to use this API is to do both of these calls. So why aren't they fused into a single function? IOW, why do I need to handle the `ReentrantLock` object at all? Why not make the unlock a side-effect of the `lockManager.releaseLock(key)` call?

Reply:
Yeah, this is reasonable.
@huangminchn Can you chime in here regarding the above question?

Reply:
Fusing two calls into one should work and we do have cases to call `lock/unlock` multiple times in the same thread, so maintaining a counter will be beneficial.
When this feature is enabled, consumer will lock a batch of keys, and to be backward compatible, the existing function will lock the same key again:
https://github.com/linkedin/venice/blob/main/clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTask.java#L225

As you mentioned, we can create a wrapper of lock to handle the reference counting logic.
So far, I don't see a case, which would call one of the functions only.

Reply:
Will optimize this in a separated PR.